_posts/2016-04-25-ryan-de-gruyter-angularjs-components.md

```diff
@@ -0,0 +1,404 @@
+
+In this article I will offer some basic guidelines on how to create a scalable AngularJS application with reusable, well encapsulated components that are easy to maintain and refactor.
+AngularJS (At the time of writing the version is 1.5.5) and its latest features offers us the ability to structure our apps as a tree of components. 
+Each component contains its own controller and template. 
+It can even have its own (relative) routing configured if you take advantage of the new component Router. 
+
+If you are on a team with multiple front-end developers you can easily divide the work by letting each developer focus on a separate component. 
+It also helps in migrating to Angular 2, though I cannot promise it will be an easy task. 
+Another bonus point is you are getting into the mindset of modern front-end development: web components.
+
+My preferred toolchain when developing AngularJS applications consists of Typescript, NPM and Webpack. 
+The sample code in this article and the sample application are created together with these tools.
+
+You can find the sample application on github: 
+[https://github.com/ryandegruyter/angularjs-components](https://github.com/ryandegruyter/angularjs-components "Sample app")
+
+
+## What is a component? ##
+
+In AngularJS a component is a directive. 
+More specifically we call it a component directive or template directive. 
+It is an approach to writing your own custom HTML elements which browsers are able to read and render. 
+HTML comes with a set of pre-defined elements, for example the `<div></div>` element or the `<span></span>` element. 
+
+By combing and nesting these standard HTML tags we can build complex UI widgets. 
+We can change their appearance and behavior dynamically with JavaScript and CSS.
+
+True web components can isolate their structure, appearance and behavior. 
+They make use of a technology called the Shadow DOM, which isolates the component in a separate DOM tree. 
+This element will have its styles and scripts encapsulated, they will not conflict with the styles and scripts inside the parent DOM. 
+
+Angular 2 takes full advantage of this technology, but unfortunately AngularJS, the framework I will be talking about in this article, does not. 
+Components you write and register inside AngularJS do not get isolated into a separate DOM tree. 
+
+Lucky for us we are able to mimic the effect of Web components by using directives. 
+We can write a reusable UI element, declare it with a custom tag and configure it by supplying attributes on the element. 
+My advice is to be sure to use correct naming conventions and a module system so styles and scripts will not conflict with each other.
+
+## Directives and components ##
+
+To create and register a custom element in AngularJS, we can use either methods:
+
+    .directive(name, factoryfunction );
```

Comment:
Since you are creating an enumeration, you should add a dash at the beginning of each line.